partridge/hmm.py

```diff
@@ -0,0 +1,256 @@
+
+                shape_spacing = to_dist - from_dist
+                stop_spacing = measures.stop_spacing[astop_idx]
+                spacing_diff = np.abs(shape_spacing - stop_spacing)
+                tprob = log_exponential_distribution(spacing_diff, beta)
+
+                trans_probs[bstop_idx][bseg_idx][astop_idx][aseg_idx] = tprob
+
+    init_probs = ddict2dict(init_probs)
+    emiss_probs = ddict2dict(emiss_probs)
+    trans_probs = ddict2dict(trans_probs)
+
+    return init_probs, emiss_probs, trans_probs
+
+
+def sparse_viterbi(N, init_probs, emiss_probs, trans_probs):
+    '''
+    Based on the implementation found here: https://git.io/fNQ6S
+    ------------------------------------------------------------
+
+    p = log probability
+    i = stop index
+    j = seg index
+
+    init_probs[i][j] = p
+    trans_probs[i1][j1][i0][j0] = p
+    emiss_probs[i][j] = p
+    '''
+    P = [{} for _ in range(N)]
+    preds = {}
+
+    for seg_idx, eprob in emiss_probs[0].items():
+        if seg_idx in init_probs[0]:
+            iprob = init_probs[0][seg_idx]
+            P[0][seg_idx] = iprob + eprob
+
+    for bstop_idx in range(1, N):
+        if bstop_idx not in emiss_probs:
+            raise NoPathError('HMM break at observation {}'.format(bstop_idx))
+
+        for bseg_idx, eprob in emiss_probs[bstop_idx].items():
+            if (
+                bstop_idx not in trans_probs or
+                bseg_idx not in trans_probs[bstop_idx]
+            ):
+                continue
+
+            for astop_idx, tprobs in trans_probs[bstop_idx][bseg_idx].items():
+                for aseg_idx, tprob in tprobs.items():
+                    if aseg_idx not in P[astop_idx]:
+                        continue
+
+                    curr_prob = P[astop_idx][aseg_idx]
+                    prob = curr_prob + eprob + tprob
+
+                    if (
+                        bseg_idx not in P[bstop_idx] or
+                        prob > P[bstop_idx][bseg_idx]
+                    ):
+                        P[bstop_idx][bseg_idx] = prob
+                        preds[bstop_idx, bseg_idx] = (astop_idx, aseg_idx)
+
+    last_stop_idx = N - 1
+
+    # Find the most probable end state
+    last_seg_idx = dict_argmax(P[-1])
+
+    path = [(last_stop_idx, last_seg_idx)]
+
+    # Reconstruct the most probable path the end state
+    for i in range(last_stop_idx):
+        if path[-1] not in preds:
+            raise NoPathError('HMM break at observation {}'.format(N - i))
+        path.append(preds[path[-1]])
+
+    path.reverse()
+
+    return path
+
+
+class CandidateCache(object):
+    def __init__(self, shape_lons, shape_lats):
+        assert shape_lons.shape[0] > 1
+        assert shape_lons.shape == shape_lats.shape
+        self.shape_lons = shape_lons
+        self.shape_lats = shape_lats
+
+    @cached_property
+    def spatial_index(self):
+        return Index((i, s.bounds, None) for i, s in enumerate(self.segments))
```

Comment:
If my understanding about how `self.segments` works - then is is fair to say that we are generating a bounds sindex for each segment of each linestring in a given route? If so, I think there is a performance opportunity being missed, wherein spatial indices could do more in terms of clustering components of the network (e.g. nodes) and clustering subsegments (single vertices) to create a higher level spatial index. 

Doing so would reduce the number of intersections that would need to be performed. I know a lot of agencies publish shapefiles that are way too detailed. These alignments/shapes would generate spatial indices (in this case) that would be equal in length to the size of the associated shape's LineString vertices length, which would risk really slow execution performance.

Reply:
> wherein spatial indices could do more in terms of clustering components of the network (e.g. nodes) and clustering subsegments (single vertices) to create a higher level spatial index.

Can you tell me more about this approach? I'm not sure I understand.

> I know a lot of agencies publish shapefiles that are way too detailed.

Earlier versions of this code used Douglas–Peucker to simplify the lines. I'm not opposed to adding an option to simplify the lines to a given precision, assuming the performance gain is meaningful and it doesn't reduce the overall accuracy. Wdyt?

Reply:
I think a line simplification would be a good first step. I'd suggest converting to a meter projection and simplifying to a ~3 meter (10 ft) resolution.

Reply:
> I'd suggest converting to a meter projection and simplifying to a ~3 meter (10 ft) resolution.

I like that idea.

Reply:
Welp, just played around with this idea for a little bit to see what gains could be made... and it's now slower. I'm probably making some dumb mistake somewhere, but wanted to share out the progress sketching this out, thus far: https://gist.github.com/kuanb/871c6df0b770f88364aff7a8abc230fc

Reply:
Oh neat! I took a quick look and I think you're more likely to see the fruits of this work in the graph building and search phases than the intersection. So there may still be hope!

Reply:
Building the spatial index might also be a tad faster.